src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -8487,24 +8486,30 @@ CheckResults StorageReplicatedMergeTree::checkData(const ASTPtr & query, Context
     else
         data_parts = getVisibleDataPartsVector(local_context);
 
-    {
-        auto part_check_lock = part_check_thread.pausePartsCheck();
+    auto part_check_lock = part_check_thread.pausePartsCheck();
+    return std::make_unique<DataValidationTasks>(std::move(data_parts), std::move(part_check_lock));
+}
+
+CheckResult StorageReplicatedMergeTree::checkDataNext(DataValidationTasksPtr & check_task_list, bool & has_nothing_to_do)
+{
 
-        for (auto & part : data_parts)
+    if (auto part = static_cast<DataValidationTasks *>(check_task_list.get())->next())
```

Comment:
Looks a little bit scary to me, when we downcast using `static_cast`. Actually here it looks safe, but still  `dynamic_cast/assert_cast/typeid_cast` would be more natural here?